### PR TITLE
Auto completion of metric tag filter for queries

### PR DESCRIPTION
**Pull request by rajatparida86**
WORK IN PROGRESS
Auto completion of metric tag filter for queries

### CODE REVIEW

File: src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java [hunk 1]
```diff
@@ -811,38 +816,22 @@ private static void logQuery(DatastoreMetricQuery query,
...
+            SplitIndex splitIndex = splitIndexDecider.decideSplitIndex(indexTags, filterTags);
+            List<String> useSplit = new ArrayList<>(splitIndex.getSplitIndexValues());
 
             long startTime = calculateRowTimeRead(query.getStartTime());
             // Use write width here, as END time is upper bound for query and end with produces the bigger timestamp
             long endTime = calculateRowTimeWrite(query.getEndTime());
             logger.info("calculated: s={} e={}", startTime, endTime);
 
-            if (useSplitField != null && !"".equals(useSplitField) && useSplitSet.size() > 0) {
+            if (splitIndex.getSplitIndexField() != null && !"".equals(splitIndex.getSplitIndexField()) && splitIndex.getSplitIndexValues().size() > 0) {
```
Comment (alexkorotkikh) [inline src/main/java/org/kairosdb/datastore/cassandra/CassandraDatastore.java]: I'd introduce a method for it, e.g. `splitIndex.isDefined()`

**Review by alexkorotkikh — CHANGES_REQUESTED** [review]
While I don't have many questions to implementation, I wonder if the feature actually is well analyzed. I afraid that kind of query modification can result in changing the behavior, which we better have to avoid.
Example (from [this graph](https://grafana.zalando.net/d/AH6tWJIZz/coast-checkout-gateway?orgId=1&refresh=1m&fullscreen&edit&panelId=66)): query `key=[timers.zmon.response.*.GET.controllers.CheckoutController.getCheckouts.m1_rate]` would be replaced with query `metric=[m1_rate]` which has a different semantic - it would collect data points for *all* Controllers and *all* methods of this application, not only for `CheckoutController::getCheckouts`, but also for `AddressController::getAddresses` and others.
I believe that's not how ZMON is expected to behave. Let's clarify that separately.

**Comment by tfrauenstein** [comment]
Editorial: We called this tuning feature 'auto completion of metric tag filter for queries' -- and we should use this name consistently also for the PR. 

More important: It is IMO too early for implementation. My expectation was that we write a short Tuning Feature Solution paper, where the major open questions we discussed are answered. From implementation design point of view, my expectation was that we do not change the core code for query processing (like proposed here), but execute query auto-completion isolated, before query processing is called. 

@alexkorotkikh With respect to your example: Yes, replacing of query "key=[timers.zmon.response.\*.GET.controllers.CheckoutController.getCheckouts.m1_rate]"  by query "metric=[m1_rate]" would change the semantics. But this was not the idea. It is 'auto completion of metric tag filter' i.e. query replaced by query "key=[timers.zmon.response.\*.GET.controllers.CheckoutController.getCheckouts.m1_rate], metric=[m1_rate]". This query should have the same semantic, but is more efficient due to metric split index usage by query processing.

**Comment by alex-on-java** [comment]
Correct me if I'm wrong, but I guess we have another PR #131, and this is not used anymore.

**Pull request closed without merging**